### ISSUE TITLE
How to ensure that OpenID configuration is finished before app startup completes?

### ISSUE THREAD

**Issue by jwillmer** [openiddict/openiddict-core#1846]
### Confirm you've already contributed to this project or that you sponsor it

- [X] I confirm I'm a sponsor or a contributor

### Version

4.x

### Question

_Hi, I ask this question for one of our developers (@MartisiusL) since I messed up the sponsorship and put it on my name instead of the company - will be fixed next month._

How can we ensure that OpenID configuration is finished and ".well-known" endpoints are available? Currently, the application starts up and accepts requests earlier than the configuration is available, which cause the error of misconfigured/inaccessible OpenID.

**Comment by kevinchalet** [comment]
Hey @jwillmer,

Thanks for sponsoring the project, much appreciated! ❤️ 

> How can we ensure that OpenID configuration is finished and ".well-known" endpoints are available? Currently, the application starts up and accepts requests earlier than the configuration is available, which cause the error of misconfigured/inaccessible OpenID.

OpenIddict - like many other libraries these days - relies on `Microsoft.Extensions.Options` to dynamically build the configuration at runtime. In the typical case, this happens only once, when the options are accessed for the first time (for OpenIddict, it means the very first HTTP request, since OpenIddict registers itself as an `IAuthenticationRequestHandler` that is invoked for each ASP.NET Core request).

This process is generally very short (even if you generate ephemeral signing/encryption keys) and shouldn't cause any timeout, as the few milliseconds/seconds it requires are way below the default timeouts enforced by the different HTTP servers offered by ASP.NET Core (Kestrel/IIS or HTTP.sys).

It's also not clear to me at all how an invalid/"misconfigured" configuration could ever be returned, even temporarily.

Do you have any additional information you can share about the symptoms you're seeing? A repro maybe?

**Comment by MartisiusL** [comment]
Hello, it is a strange behavior for us as well. 

Basically all what you said is probably true, we experienced it that it only gets loaded with the very first HTTP request and if that request is delayed or similar then it works fine. However, this can sometimes take even a minute or more which make the delay time not a reasonable solution.

Our environment is dependent on this project and we have multiple applications that at any given time if this application restarts tries to reconnect as soon as possible. This cause the problem, since as described, if the OpenID configuration is not ready it fails. However, it not only fails, but during the first request, it tries to configure the well-known endpoints and caches them without verifying that they work. And for some reason after they are cached, if this caching happens before it is actually loaded it gets lost and all the ".well-known" endpoints are never available unless again the application restarts.

For sometime we had a solution that helped, where use middleware to decline all requests until well-known is configured by just trying to access it and validate. However, this solution fails when we try to scale out and now we are here.

So my question was how to validate without http request to self, that the setup was done. I will add some code snippets on the configuration.

```
public static IServiceCollection ConfigureOpenIddict(this IServiceCollection services, AuthorizationServerSettings authorizationServerSettings)
      {
          services.AddOpenIddict()
              .AddCore(options =>
              {
                  options.UseMongoDb()
                      .ReplaceDefaultAuthorizationEntity<DAL.Models.OpenIddict.OpenIddictAuthorization>()
                      .ReplaceDefaultApplicationEntity<DAL.Models.OpenIddict.OpenIddictApplication>()
                      .ReplaceDefaultScopeEntity<DAL.Models.OpenIddict.OpenIddictScope>()
                      .ReplaceDefaultTokenEntity<DAL.Models.OpenIddict.OpenIddictToken>();
                  options.DisableEntityCaching();
              })
              .AddServer(options =>
              {
                  options.AllowAuthorizationCodeFlow()
                      .AllowRefreshTokenFlow()
                      .UseReferenceRefreshTokens()
                      .SetRefreshTokenLifetime(TimeSpan.FromSeconds(authorizationServerSettings.RefreshTokenLifeTimeInSeconds));
                  options.RegisterScopes(Scopes.Email, Scopes.Profile, Scopes.Roles, Scopes.OfflineAccess);
                  options.DisableAccessTokenEncryption();
                options.UseAspNetCore()
                    .EnableTokenEndpointPassthrough()
                    .EnableAuthorizationEndpointPassthrough()
                    .EnableUserinfoEndpointPassthrough()
                    .EnableLogoutEndpointPassthrough();
                options.RemoveEventHandler(ValidateAuthorizedParty.Descriptor);
options.SetRefreshTokenReuseLeeway(TimeSpan.FromSeconds(authorizationServerSettings.RefreshTokenReuseLeewayInSeconds));
            })
        .AddValidation(options =>
        {
            options.UseLocalServer();
            options.UseAspNetCore();
        });
        services.AddSingleton<IOptionsMonitor<OpenIddictServerOptions>, OpenIddictServerOptionsProvider>();
        services.AddSingleton<IConfigureOptions<OpenIddictServerOptions>, OpenIddictServerOptionsInitializer>();
        services.AddSingleton<IOptionsMonitor<OpenIddictMongoDbOptions>, OpenIddictMongoDbOptionsProvider>();
        services.AddSingleton<IConfigureOptions<OpenIddictMongoDbOptions>, OpenIddictMongoDbOptionsInitializer>();
        return services;
    }
```


```
public void Configure(OpenIddictServerOptions options)
    {
        var tenant = _oAuthTenantProvider.GetCurrentTenant();

        ILifetimeScope tenantScope = _multitenantContainer.GetTenantScope(tenant.Name);
        ITokenSigningSecurityKeyManager tokenSigningSecurityKeyManager = tenantScope.Resolve<ITokenSigningSecurityKeyManager>();
        var signingKey = tokenSigningSecurityKeyManager.Get();
        var rsaCredentials = new SigningCredentials(signingKey, SecurityAlgorithms.RsaSha512);
        options.SigningCredentials.Add(rsaCredentials);

        var securityKey = tenant.ConnectionSecurityKey;
        var key = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(securityKey));
        var symmetricCredentials = new SigningCredentials(key, SecurityAlgorithms.HmacSha256);
        options.SigningCredentials.Add(symmetricCredentials);

        var encryptionKey = new RsaSecurityKey(RSA.Create(keySizeInBits: 2048));
        options.EncryptionCredentials.Add(new(encryptionKey, SecurityAlgorithms.RsaOAEP, SecurityAlgorithms.Aes256CbcHmacSha512));
        options.DisableAccessTokenEncryption = true;
        string issuerUrl = tenant.Applications[Shared.DTO.Enums.KnownApplication.UserManagement] + tenant.Name;
        options.AuthorizationEndpointUris.Clear();
        options.AuthorizationEndpointUris.Add(new Uri($"{issuerUrl}/v2/connect/authorize", UriKind.Absolute));
        options.LogoutEndpointUris.Clear();
        options.LogoutEndpointUris.Add(new Uri($"{issuerUrl}/v2/connect/logout", UriKind.Absolute));
        options.TokenEndpointUris.Clear();
        options.TokenEndpointUris.Add(new Uri($"{issuerUrl}/v2/connect/token", UriKind.Absolute));
        options.UserinfoEndpointUris.Clear();
        options.UserinfoEndpointUris.Add(new Uri($"{issuerUrl}/v2/connect/userinfo", UriKind.Absolute));
        options.CryptographyEndpointUris.Clear();
        options.CryptographyEndpointUris.Add(new Uri($"{issuerUrl}/.well-known/jwks", UriKind.Absolute));
        options.ConfigurationEndpointUris.Clear();
        options.ConfigurationEndpointUris.Add(new Uri($"{issuerUrl}/.well-known/openid-configuration", UriKind.Absolute));
        options.Issuer = new Uri(issuerUrl);
        options.AccessTokenLifetime = TimeSpan.FromSeconds(_authorizationServerSettings.AccessTokenLifeTimeInSeconds);
    }
```

These are two main points that this gets configured.

**Comment by kevinchalet** [comment]
> And for some reason after they are cached, if this caching happens before it is actually loaded it gets lost and all the ".well-known" endpoints are never available unless again the application restarts.
For sometime we had a solution that helped, where use middleware to decline all requests until well-known is configured by just trying to access it and validate. However, this solution fails when we try to scale out and now we are here.

It's definitely not a normal behavior. I see you use a custom `IOptionsMonitor<OpenIddictServerOptions>`: are you sure there's not a bug there? Can you share the implementation?

> So my question was how to validate without http request to self, that the setup was done. I will add some code snippets on the configuration.

You could resolve `IOptionsMonitor<OpenIddictServerOptions>` from the DI and access `.CurrentValue`: if no exception is thrown, the options are valid.

**Comment by MartisiusL** [comment]
That's our implementation of custom class:

```
public class OpenIddictServerOptionsProvider : IOptionsMonitor<OpenIddictServerOptions>
{
    private readonly ConcurrentDictionary<(string Name, string Tenant), Lazy<OpenIddictServerOptions>> _cache;
    private readonly IOptionsFactory<OpenIddictServerOptions> _optionsFactory;
    private readonly IOAuthTenantProvider _oAuthTenantProvider;

    /// <summary>
    /// ctor
    /// </summary>
    /// <param name="optionsFactory"></param>
    /// <param name="oAuthTenantProvider"></param>
    public OpenIddictServerOptionsProvider(
        IOptionsFactory<OpenIddictServerOptions> optionsFactory,
        IOAuthTenantProvider oAuthTenantProvider)
    {
        _cache = new ConcurrentDictionary<(string, string), Lazy<OpenIddictServerOptions>>();
        _optionsFactory = optionsFactory;
        _oAuthTenantProvider = oAuthTenantProvider;
    }

    /// <summary>
    /// CurrentValue
    /// </summary>
    public OpenIddictServerOptions CurrentValue => Get(Microsoft.Extensions.Options.Options.DefaultName);

    /// <summary>
    /// Will generate key on first execution and then will store it in cache.
    /// </summary>
    /// <param name="name"></param>
    /// <returns></returns>
    public OpenIddictServerOptions Get(string name)
    {
        string tenantName;
        try
        {
            var tenant = _oAuthTenantProvider.GetCurrentTenant();
            tenantName = tenant.Name;
        }
        catch (InvalidOperationException)
        {
            tenantName = null;
        }

        Lazy<OpenIddictServerOptions> Create() => new(() => _optionsFactory.Create(name));
        return _cache.GetOrAdd((name, tenantName), _ => Create()).Value;
    }

    /// <summary>
    /// OnChage
    /// </summary>
    /// <param name="listener"></param>
    /// <returns></returns>
    public IDisposable OnChange(Action<OpenIddictServerOptions, string> listener) => null;
}
```

and I will try to access it through DI and see what happens.

**Comment by MartisiusL** [comment]
So accessing it through DI results in the same problem, with my simple test with such a middleware:

```
            var openIdConfiguration = context.RequestServices.GetService<IOptions<OpenIddictServerOptions>>();
            try
            {
                var optionsMonitor = context.RequestServices.GetService<IOptionsMonitor<OpenIddictServerOptions>>();
                var optionsValue = optionsMonitor.CurrentValue;
            }
            catch(Exception)
            {
                Log.Debug("Options were not available");
            }
```

It does never go into exception, however, it tries to setup the endpoints earlier and cause the well-known endpoints to end up being inaccessible.

Some error context for reference

```
fail: Microsoft.AspNetCore.Server.Kestrel[13]
      Connection id "0HMSDSEMFOKIE", Request id "0HMSDSEMFOKIE:00000001": An unhandled exception was thrown by the application.
      System.InvalidOperationException: IDX20803: Unable to obtain configuration from: 'https://localhost:44372/test/.well-known/openid-configuration'.
         at Microsoft.IdentityModel.Protocols.ConfigurationManager`1.GetConfigurationAsync(CancellationToken cancel)
         at Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectHandler.HandleChallengeAsyncInternal(AuthenticationProperties properties)
         at Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectHandler.HandleChallengeAsync(AuthenticationProperties properties)
         at Microsoft.AspNetCore.Authentication.AuthenticationHandler`1.ChallengeAsync(AuthenticationProperties properties)
         at Microsoft.AspNetCore.Authentication.AuthenticationService.ChallengeAsync(HttpContext context, String scheme, AuthenticationProperties properties)
         at Microsoft.AspNetCore.Authorization.Policy.AuthorizationMiddlewareResultHandler.HandleAsync(RequestDelegate next, HttpContext context, AuthorizationPolicy policy, PolicyAuthorizationResult authorizeResult)
         at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
         at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
         at Maranics.UserManagement.Core.Tools.Helpers.ResponseHeadersConfigurationHelper.<>c.<<DisablePdfAndFlashCrossDomainRequests>b__1_0>d.MoveNext() in C:\src\maranics\UserManagementApp\Maranics.UserManagement\Maranics.UserM
anagement.Core\Tools\Helpers\ResponseHeadersConfigurationHelper.cs:line 27
```

**Comment by MartisiusL** [comment]
And the way I reproduce it locally is quite stupid, but works, is that with this setup and well quite heavy application while it is starting I just keep spamming refresh button on the "/.well-known/openid-configuration" endpoint until it fully loads up. And it gets into inaccessible. If I wait for some time without spamming, everything goes smoothly, and it works without any problems.

**Comment by kevinchalet** [comment]
What happens when you visit `https://localhost:44372/test/.well-known/openid-configuration` in your browser? Do you see an exception logged in the server logs? Is an HTTP error response returned?

(in any case, it's very likely caused by your specific configuration/environment, as it's the first time I get such a report)

**Comment by MartisiusL** [comment]
Yes, in the server we get the error I pasted as context for reference. And browser is just an empty screen.

**Comment by kevinchalet** [comment]
> Yes, in the server we get the error I pasted as context for reference. And browser is just an empty screen.

No, the log you shared are the client logs, I need the logs of the authorization server 😃 

Is there a chance you could build a small repro so I can give it a try locally?

**Comment by MartisiusL** [comment]
Okay, so I tried really hard to build a small reproducable project, with no success really, but during which I most likely found the problem or the area that causes that. So we basically have one HostedService that does the cleanup for unused or so application registrations:

```
public class OpenIddictApplicationSync : IHostedService
{
    private readonly IOAuthTenantProvider _oAuthTenantProvider;
    private readonly IConfigurationService _configurationService;
    private readonly IOIddictApplicationManager _oIddictApplication;

    public OpenIddictApplicationSync(IOAuthTenantProvider oAuthTenantProvider, IConfigurationService configurationService, IOIddictApplicationManager oIddictApplication)
    {
        _configurationService = configurationService;
        _oAuthTenantProvider = oAuthTenantProvider;
        _oIddictApplication = oIddictApplication;
    }

    public async Task StartAsync(CancellationToken cancellationToken)
    {
        var tenants = _configurationService.GetTenants<TenantConfiguration>();
        foreach (TenantConfiguration tenant in tenants)
        {
            // Since there is no http context, the tenant cannot be determined by normal ways, so we override it with the tenant name that we have here
            _oAuthTenantProvider.TenantOverride = tenant.Name;

            await _oIddictApplication.SyncAuthApplicationsForTenant(tenant, cancellationToken);
        }

        _oAuthTenantProvider.TenantOverride = null;
    }

    public Task StopAsync(CancellationToken cancellationToken) => Task.CompletedTask;
}
```

Now that "sync" function gets some entries from mongoDb, compares it with what openIdManager has and deletes accordingly in some manner:

```
    public async Task SyncAuthApplicationsForTenant(TenantConfiguration tenantConfiguration, CancellationToken cancellationToken)
    {
        var dataContext = new DataContext();
        dataContext.Init(tenantConfiguration.MongoConnection);
        var authApps = (await dataContext.OAuthApplications.AsQueryable().ToListAsync(cancellationToken))
            .Where(x => x.State != State.Removed)
            .Select(x => x.ToDto());

        var existingApps = (await GetAll()).ToList();
        foreach (var openIddictApplication in existingApps)
        {
            if (!authApps.Any(x => x.ClientId == openIddictApplication.ClientId))
            {
                await DeleteApplication(openIddictApplication.ClientId);
            }
        }

        foreach (var application in authApps)
        {
            var existingApp = existingApps.FirstOrDefault(x => x.ClientId == application.ClientId);
            if (existingApp != null)
            {
                IEnumerable<string> postLogoutRedirectUris = application.PostLogoutRedirectUrls ?? Enumerable.Empty<string>();
                existingApp.RedirectUris = new ReadOnlyCollection<string>(application.RedirectUrls.ToList());
                existingApp.PostLogoutRedirectUris = new ReadOnlyCollection<string>(postLogoutRedirectUris.ToList());
                await _openIddictApplicationManager.UpdateAsync(existingApp, cancellationToken);
                continue;
            }

            var newOpenIdDictApplicationDescriptor = new OpenIddictApplicationDescriptor
            {
                ClientId = application.ClientId,
                ClientSecret = application.ClientSecret,
                DisplayName = application.Name
            };
            newOpenIdDictApplicationDescriptor.Permissions.AddRange(GetPermissions());
            newOpenIdDictApplicationDescriptor.RedirectUris.AddRange(application.RedirectUrls.Select(x => new Uri(x)));
            if (application.PostLogoutRedirectUrls != null)
            {
                newOpenIdDictApplicationDescriptor.PostLogoutRedirectUris.AddRange(application.PostLogoutRedirectUrls.Select(x => new Uri(x)));
            }

            await _openIddictApplicationManager.CreateAsync(newOpenIdDictApplicationDescriptor, cancellationToken);
        }

        var resources = existingApps.Select(x => x.ClientId);
        await AssociateScopeWithResources(OpenIddictConstants.Scopes.Profile, resources, true);
        await AssociateScopeWithResources(OpenIddictConstants.Scopes.Email, resources, true);
        await AssociateScopeWithResources(ScopeConstants.User, resources, true);
    }
```

I was trying to get into the same cycle, with no success sadly, but removing this service at all helps and the application no longer causes this behavior. As a temporary measure, I added 2 minutes delay before running that service. However, after doing so, I ran into a different problem, maybe it is related, but now all the well-known's are configured correctly, but trying to authenticate results in this error:

```
fail: Microsoft.AspNetCore.Server.Kestrel[13]
      Connection id "0HMSEJQKKJGRK", Request id "0HMSEJQKKJGRK:00000014": An unhandled exception was thrown by the application.
      System.Exception: An error was encountered while handling the remote login.
       ---> Microsoft.IdentityModel.Protocols.OpenIdConnect.OpenIdConnectProtocolException: Message contains error: 'invalid_grant', error_description: 'The specified token is invalid.', error_uri: 'https://documentation.openiddict.com/errors/ID2004'.
         at Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectHandler.RedeemAuthorizationCodeAsync(OpenIdConnectMessage tokenEndpointRequest)
         at Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectHandler.HandleRemoteAuthenticateAsync()
         --- End of inner exception stack trace ---
         at Microsoft.AspNetCore.Authentication.RemoteAuthenticationHandler`1.HandleRequestAsync()
         at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
         at Maranics.UserManagement.Core.Tools.Helpers.ResponseHeadersConfigurationHelper.<>c.<<DisablePdfAndFlashCrossDomainRequests>b__1_0>d.MoveNext() in /app/Maranics.UserManagement.Core/Tools/Helpers/ResponseHeadersConfigurationHelper.cs:line 27
      --- End of stack trace from previous location ---
         at Maranics.UserManagement.Core.Tools.Helpers.ResponseHeadersConfigurationHelper.<>c.<<UseNosniffContentTypeHeader>b__0_0>d.MoveNext() in /app/Maranics.UserManagement.Core/Tools/Helpers/ResponseHeadersConfigurationHelper.cs:line 16
      --- End of stack trace from previous location ---
         at Joonasw.AspNetCore.SecurityHeaders.ReferrerPolicy.ReferrerPolicyMiddleware.Invoke(HttpContext context)
         at Swashbuckle.AspNetCore.SwaggerUI.SwaggerUIMiddleware.Invoke(HttpContext httpContext)
         at Swashbuckle.AspNetCore.Swagger.SwaggerMiddleware.Invoke(HttpContext httpContext, ISwaggerProvider swaggerProvider)
         at Microsoft.AspNetCore.ResponseCompression.ResponseCompressionMiddleware.InvokeCore(HttpContext context)
         at Maranics.UserManagement.Core.Tools.Middleware.TenantProvisionCheckMiddleware.InvokeAsync(HttpContext context) in /app/Maranics.UserManagement.Core/Tools/Middleware/TenantProvisionCheckMiddleware.cs:line 52
         at Autofac.Integration.AspNetCore.Multitenant.MultitenantRequestServicesMiddleware.Invoke(HttpContext context)
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)
```

Do you have an idea why, is it related or similar?

**Comment by MartisiusL** [comment]
So it only happens when I run multiple instances of kubernetes pods. But we have a persistence key storage setup:

```
string redisConnectionString = _configuration.GetConnectionString(RedisConnectionStringName);
var redis = ConnectionMultiplexer.Connect(redisConnectionString);
services.AddDataProtection()
    .PersistKeysToStackExchangeRedis(redis, "DataProtection-Keys").SetApplicationName("SSO");
```

**Comment by kevinchalet** [comment]
> Do you have an idea why, is it related or similar?

It's really hard for me to make any guess without seeing any authorization server logs and without having access to the complete code 😅 

> So it only happens when I run multiple instances of kubernetes pods. But we have a persistence key storage setup:

By default, OpenIddict doesn't use ASP.NET Core Data Protection to protect its tokens (it's a separate package/opt-in integration). In this case, what matters is the signing and encryption credentials and AFAICT by looking at the snippets you shared, you only sync the signing credentials but not the encryption credentials that are generated dynamically and not persisted:

```csharp
var encryptionKey = new RsaSecurityKey(RSA.Create(keySizeInBits: 2048));
options.EncryptionCredentials.Add(new(encryptionKey, SecurityAlgorithms.RsaOAEP, SecurityAlgorithms.Aes256CbcHmacSha512));
```

You'll likely need to adopt a similar pattern as the one you're using for signing credentials with `ITokenSigningSecurityKeyManager`.

**Comment by MartisiusL** [comment]
Well your guess might be exactly what's missing, since I was referring to this discussion: https://github.com/openiddict/openiddict-core/issues/607 that is quite old now and things might have changed. So we definitely don't have a instance sharing pattern for encryption key. Is there a chance that you have any example of how it's done in other projects?

**Comment by kevinchalet** [comment]
> Is there a chance that you have any example of how it's done in other projects?

It actually works exactly like signing credentials: you must generate a key and share it between all the instances of your authorization server. It can be a RSA key (asymmetric) as in your current snippet (but you must store it and resolve it instead of generating a new one every time the app starts) or an AES key (symmetric) as in this sample: https://github.com/openiddict/openiddict-samples/blob/eb5cc206852060c194b46d07363deb4016cc27f1/samples/Zirku/Zirku.Server/Program.cs#L70-L77

You can also generate and use a X.509 certificate containing a RSA key, as shown here: https://documentation.openiddict.com/configuration/encryption-and-signing-credentials.html#registering-a-certificate-recommended-for-production-ready-scenarios.

**Comment by MartisiusL** [comment]
Hello again, so just writing with final confirmation that we resolved all out issues and it works now. Thank you very much for your quick and informative help.

**Final resolution of the first problem:**
Our problem is our very custom implementation. Our project is quite tightly coupled with OpenIdDict in a way, that we reuse OpenIdDict `applications` table for ourselves. In order to do so, we have to sync it. This synchronization at the start of the application was causing issues, because of the race condition. We have multiple `.well-known` configurations, separate for each tenant and basically some of them were getting delete request directly to OpenId and therefore they were missing afterwards.
I resolved this by just delaying the sync until the configuration is ready and now it works as expected.
One additional thing I noticed during testing and trying to reproduce, I used the project that is given as sample: [Velusia](https://github.com/openiddict/openiddict-samples/tree/dev/samples/Velusia) and if you spam the `https://localhost:44313/.well-known/openid-configuration` you still get the error in the project logs of 
```
fail: Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware[1]
      An unhandled exception has occurred while executing the request.
      System.Threading.Tasks.TaskCanceledException: A task was canceled.
         at OpenIddict.Server.AspNetCore.OpenIddictServerAspNetCoreHandlers.ProcessJsonResponse`1.HandleAsync(TContext context)
         at OpenIddict.Server.OpenIddictServerDispatcher.DispatchAsync[TContext](TContext context)
         at OpenIddict.Server.OpenIddictServerDispatcher.DispatchAsync[TContext](TContext context)
         at OpenIddict.Server.OpenIddictServerHandlers.Discovery.ApplyConfigurationResponse`1.HandleAsync(TContext context)
         at OpenIddict.Server.OpenIddictServerDispatcher.DispatchAsync[TContext](TContext context)
         at OpenIddict.Server.OpenIddictServerDispatcher.DispatchAsync[TContext](TContext context)
         at OpenIddict.Server.AspNetCore.OpenIddictServerAspNetCoreHandler.HandleRequestAsync()
         at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
         at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddlewareImpl.Invoke(HttpContext context)
```
Maybe that's by design, but actually this cancellation was visible in our logs as well, which may also be sort of the reason why it was getting delayed or similar.

**Resolution for the second problem:**
Basically all that I needed to know was that the key to be shared is `RsaSecurityKey` and not the "ASP.NET Core Data Protection". After knowing this, I was not waiting for your answer and implemented my own version of sharing, maybe not the greatest, but works for now, I'm just taking entire key, serializing it and deserializing it and saving to internal Redis instance. Works for now, and we may update it to your suggested versions that sound even simplier in some way.

That's it. Thank you again for your time and help!

**Comment by jwillmer** [comment]
@kevinchalet thank you for your support. It saved us a lot of time and was very helpful and informative ❤️

**Comment by kevinchalet** [comment]
Glad you were able to make your scenario work! 👍🏻 

> Maybe that's by design, but actually this cancellation was visible in our logs as well, which may also be sort of the reason why it was getting delayed or similar.

It is, actually: when you hit F5 super hard in your browser, the refresh rate can become so high that the browser ends up canceling some of the HTTP requests, which manifests as a TCP RST that is detected by ASP.NET Core and propagated via `HttpContext.RequestAborted`: since OpenIddict uses that cancellation token when writing the response, an exception is thrown when the request is marked as aborted.

If it bothers you, you can write an ASP.NET Core middleware or exception handler to ignore `OperationCanceledException`s/`TaskCanceledException`s

> That's it. Thank you again for your time and help!

> @kevinchalet thank you for your support. It saved us a lot of time and was very helpful and informative ❤️

My pleasure! Thanks again for sponsoring the project 👍🏻
